[lldb][progress][NFC] Add unit test for progress reports

chelcassanova opened the pull request:
This test is being added as a way to check the behaviour of how progress events are broadcasted when reports are started and ended with the current implementation of progress reports. Here we're mainly checking and ensuring that the current behaviour is that progress events are broadcasted individually and placed in the event queue in order of their creation.

chelcassanova committed:
[lldb][progress][NFC] Add unit test for progress reports

This test is being added as a way to check the behaviour of how progress
events are broadcasted when reports are started and ended with the
current implementation of progress reports. Here we're mainly checking
and ensuring that the current behaviour is that progress events are
broadcasted individually and placed in the event queue in order of
their creation.

JDevlieghere commented on lldb/unittests/Core/ProgressReportTest.cpp:
```diff
@@ -0,0 +1,105 @@
+namespace {
+class ProgressReportTest : public ::testing::Test {
+public:
+  void SetUp() override {
+    FileSystem::Initialize();
+    HostInfo::Initialize();
+    PlatformMacOSX::Initialize();
+    Debugger::Initialize(nullptr);
+  }
+  void TearDown() override {
+    Debugger::Terminate();
+    PlatformMacOSX::Terminate();
+    HostInfo::Terminate();
+    FileSystem::Terminate();
+  }
```
There's a utility that simplifies this called `SubsystemRAII`. Here's how you use that:

```
#include "TestingSupport/SubsystemRAII.h"

class ProgressReportTest : public ::testing::Test {
SubsystemRAII<FileSystem, HostInfo, PlatformMacOSX, Debugger> subsystems;
}
```

chelcassanova replied:
Lovely, I'll add this in here

JDevlieghere commented on lldb/unittests/Core/ProgressReportTest.cpp:
```diff
@@ -0,0 +1,105 @@
+  }
+};
+} // namespace
+TEST_F(ProgressReportTest, TestReportCreation) {
```
Nit: missing newline

JDevlieghere commented on lldb/unittests/Core/ProgressReportTest.cpp:
```diff
@@ -0,0 +1,105 @@
+  // Progress report objects should be destroyed at this point so
+  // get each report from the queue and check that they've been
+  // destroyed in reverse order
+  std::this_thread::sleep_for(timeout);
```
This shouldn't be necessary: you've done the progress reporting from the same thread that you're getting the event from so since this is linear code, all events should be queued up in the broadcaster.

chelcassanova replied:
Ah, I thought I had to give a tiny bit of time for the objects to be destroyed, I'll remove the timeouts here then.

medismailben commented:
This is only testing the `Progress` class constructor behavior. Could you check that the `Progress::Increment` method works as expected ?

bulbazord commented on lldb/unittests/Core/ProgressReportTest.cpp:
```diff
@@ -0,0 +1,105 @@
+#include "Plugins/Platform/MacOSX/PlatformMacOSX.h"
```
Don't forget to include the license header in this file

chelcassanova replied:
Will do! I told myself I'd add it at the end then I fully forgot to do that 🙃

clayborg commented on lldb/unittests/Core/ProgressReportTest.cpp:
```diff
@@ -0,0 +1,105 @@
+  // Create progress reports and check that their respective events for having
+  // started are broadcasted
+  {
+    Progress progress1("Progress report 1", "Starting report 1");
+    EXPECT_TRUE(listener_sp->GetEvent(event_sp, timeout));
+
+    data = ProgressEventData::GetEventDataFromEvent(event_sp.get());
+    ASSERT_EQ(data->GetDetails(), "Starting report 1");
+
+    Progress progress2("Progress report 2", "Starting report 2");
+    EXPECT_TRUE(listener_sp->GetEvent(event_sp, timeout));
+
+    data = ProgressEventData::GetEventDataFromEvent(event_sp.get());
+    ASSERT_EQ(data->GetDetails(), "Starting report 2");
+
+    Progress progress3("Progress report 3", "Starting report 3");
+    EXPECT_TRUE(listener_sp->GetEvent(event_sp, timeout));
+    ASSERT_TRUE(event_sp);
+
+    data = ProgressEventData::GetEventDataFromEvent(event_sp.get());
+    ASSERT_EQ(data->GetDetails(), "Starting report 3");
```
If we are looking to verify things are delivered in the order that they were supplied, do you want to do all of the event checking outside of this loop? Otherwise the only thing on the event queue here is the data we are looking for so this doesn't help us to verify that we got things in the order that they were pushed onto the queue. We should be able to do:
```
{
  Progress progress1("Progress report 1", "Starting report 1");
  Progress progress2("Progress report 2", "Starting report 2");
  Progress progress3("Progress report 3", "Starting report 3");
}
// Now check that we got all events in the right order for both the start and completed
// check progress started for "Progress report 1"
// check progress started for "Progress report 2"
// check progress started for "Progress report 3"
// check progress completed for "Progress report 1"
// check progress completed for "Progress report 2"
// check progress completed for "Progress report 3"
```
This will help us verify the order of things. 

Might be a good idea to do a check for progress reports with a valid total as well, and test that if we increment too many times that we don't deliver multiple completed events
```
{
  Progress progress1("Progress report 1", "Starting report 1");
  Progress progress2("Progress report 2", "Starting report 2");
  Progress progress3("Progress report 3", "Starting report 3");
  Progress progress4("Progress report 4", "Starting report 4", 2);
  progress4.Increment(1, "progress detail 1");
  // This should cause the progress to complete
  progress4.Increment(1, "progress detail 2"); 
  // The progress is already completed, I would expect to not see any more events for this progress
  progress4.Increment(1, "too many, we shouldn't see this!");
}
```
And when verifying the events for `progress4`:
```
ASSERT_TRUE(data->IsFinite());
ASSERT_FALSE|ASSERT_TRUE(data->GetCompleted()); // Make sure this is correct for each event
ASSERT_EQ(data->GetTotal(), 2);
ASSERT_EQ(data->GetMessage(), "..."); // Make sure this is correct for each increment
```

chelcassanova committed:
Simplify set up, remove timeouts, add license

Uses `SubsystemRAII` to initialize the necessary platform and host
information. Since the debugger cannot be initialized with no
arguments (which is how `SubsystemRAII` will initialize everything), the
debugger will be initialized the usual way. Also removes the thread
timeouts and adds the LLVM license header.

JDevlieghere commented:
> This is only testing the `Progress` class constructor behavior. Could you check that the `Progress::Increment` method works as expected ?

While it would be nice to have full test coverage, the goal was to cover the existing behavior which we'll modify/extend to coalesce events as discussed in the [RFC](https://discourse.llvm.org/t/rfc-improve-lldb-progress-reporting/75717/). Increments are orthogonal to that so I think it's fine to keep that for a separate PR.

clayborg commented on lldb/unittests/Core/ProgressReportTest.cpp:
```diff
@@ -0,0 +1,105 @@
+    EXPECT_TRUE(listener_sp->GetEvent(event_sp, timeout));
+
+    data = ProgressEventData::GetEventDataFromEvent(event_sp.get());
+    ASSERT_EQ(data->GetDetails(), "Starting report 1");
```
It would be good to verify a few more things here for each start event:
```
ASSERT_FALSE(data->IsFinite());
ASSERT_FALSE(data->GetCompleted());
ASSERT_EQ(data->GetTotal(), 1);
ASSERT_EQ(data->GetMessage(), "Progress report 1: Starting report 1");
```

clayborg commented on lldb/unittests/Core/ProgressReportTest.cpp:
```diff
@@ -0,0 +1,105 @@
+  data = ProgressEventData::GetEventDataFromEvent(event_sp.get());
+
+  ASSERT_EQ(data->GetTitle(), "Progress report 3");
+  ASSERT_TRUE(data->GetCompleted());
```
Add a few more tests for each complete event
```
ASSERT_FALSE(data->IsFinite());
ASSERT_EQ(data->GetMessage(), "...");
```

clayborg reviewed (commented):
It is great to have unit tests for this. I remember with your previous patch we caught an issue where the completed event might not be delivered, or it would be delivered without the values being set correctly, so this will be great to be able to catch this kind of stuff when we make changes.

chelcassanova committed:
Check all events outside of scope, use correct check for total

Checks that the progress events were received in order of report
creation and deletion outside of the scope that the events were created
in, also checks for more information from the progress event data.

`IsFinite()` from DebuggerEvents would return true if the total was not
UINT64_MAX. Since we no longer use this to specify that an event has no
total this has been changed to check that the total is not 1.

chelcassanova commented on lldb/include/lldb/Core/DebuggerEvents.h:
```diff
@@ -39,7 +39,7 @@ class ProgressEventData : public EventData {
 
   uint64_t GetID() const { return m_id; }
-  bool IsFinite() const { return m_total != UINT64_MAX; }
+  bool IsFinite() const { return m_total != 1; }
```
I have this here but let me know if it's better to have this in its own PR

clayborg replied:
I mention creating a variable in the Progress class in an inline comment below:
```
bool IsFinite() const { return m_total != Progress::NO_TOTAL; }
```

clayborg commented on lldb/unittests/Core/ProgressReportTest.cpp:
```diff
@@ -0,0 +1,125 @@
+
+TEST_F(ProgressReportTest, TestReportCreation) {
+  std::chrono::milliseconds timeout(100);
+  const unsigned long long NO_TOTAL = 1;
```
We should define this in Progress class:
```
class Progress {
constexpr uint64_t NO_TOTAL = UINT64_MAX;
```
UINT64_MAX is better than 1 because someone might actually create a valid finite progress with only 1 item.

chelcassanova replied:
I can use the changes from https://github.com/llvm/llvm-project/pull/79912 in here for checking the total progress.

clayborg commented on lldb/unittests/Core/ProgressReportTest.cpp:
```diff
@@ -0,0 +1,125 @@
+  ASSERT_EQ(data->GetDetails(), "Starting report 1");
+  ASSERT_FALSE(data->IsFinite());
+  ASSERT_FALSE(data->GetCompleted());
+  ASSERT_EQ(data->GetTotal(), NO_TOTAL);
```
```
ASSERT_EQ(data->GetTotal(), Progress::NO_TOTAL);
```

chelcassanova committed:
Use static var to check progress total

Use the changes implemented in
https://github.com/llvm/llvm-project/pull/79912 to use a static variable
from `Progress` to ensure that the report was created with a
non-deterministic total.

JDevlieghere reviewed (approved):
There's a handful of comments that are missing a period. Otherwise this LGTM.
